### PR TITLE
CodeCommit approval rule template association

### LINKED ISSUE #11461: Add resource for CodeCommit approval rule templates
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS recently added approval rule templates to CodeCommit to improve the pull request process. Need to add a resource for this.

### New or Affected Resource(s)

* aws_codecommit_approval_rule_template

### References

* https://docs.aws.amazon.com/codecommit/latest/APIReference/API_CreateApprovalRuleTemplate.html
* https://docs.aws.amazon.com/codecommit/latest/APIReference/API_GetApprovalRuleTemplate.html

### PR DESCRIPTION
**Pull request by carthewd**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #11461

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```
New Resource: aws_codecommit_approval_rule_association
```
### Example
```hcl
resource "aws_codecommit_approval_rule_association" "test" {
    template_name = "test2-rule"
    repository_names = ["test1", "test2"]
}
```
Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSCodeCommitApprovalRuleAssociation'                                                                  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCodeCommitApprovalRuleAssociation -timeout 120m
=== RUN   TestAccAWSCodeCommitApprovalRuleAssociation_basic
=== PAUSE TestAccAWSCodeCommitApprovalRuleAssociation_basic
=== RUN   TestAccAWSCodeCommitApprovalRuleAssociation_withChanges
=== PAUSE TestAccAWSCodeCommitApprovalRuleAssociation_withChanges
=== CONT  TestAccAWSCodeCommitApprovalRuleAssociation_basic
=== CONT  TestAccAWSCodeCommitApprovalRuleAssociation_withChanges
--- PASS: TestAccAWSCodeCommitApprovalRuleAssociation_basic (41.50s)
--- PASS: TestAccAWSCodeCommitApprovalRuleAssociation_withChanges (82.65s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       82.691s

```

### CODE REVIEW

**Comment by xenji** [comment]
Hey folks, any news when this will hit the release train? Would be highly appreciated!

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by nickjmv** [comment]
Hi, any idea when this will be merged?

File: aws/resource_aws_codecommit_approval_rule_association.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		Create: resourceAwsCodeCommitApprovalRuleAssociationCreate,
+		Update: resourceAwsCodeCommitApprovalRuleAssociationUpdate,
+		Read:   resourceAwsCodeCommitApprovalRuleAssociationRead,
+		Delete: resourceAwsCodeCommitApprovalRuleAssociationDelete,
+
+		Schema: map[string]*schema.Schema{
+			"template_name": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
```
Comment (DrFaust92) [inline aws/resource_aws_codecommit_approval_rule_association.go:23]: according to docs lets add length validation https://docs.aws.amazon.com/codecommit/latest/APIReference/API_AssociateApprovalRuleTemplateWithRepository.html

File: aws/resource_aws_codecommit_approval_rule_association.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+			},
+			"repository_names": {
+				Type:     schema.TypeSet,
+				Elem:     &schema.Schema{Type: schema.TypeString},
```
Comment (DrFaust92) [inline aws/resource_aws_codecommit_approval_rule_association.go:27]: i think we should stick to api and have this be a single value and users can use `for_each` if needed

**Review by DrFaust92 — CHANGES_REQUESTED** [review]
Hi @carthewd, thanks for the contribution. some comments + take a look at https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/CONTRIBUTING.md for for more things like using `finder` functions and adding `disappears` test

**Comment by johngmyers** [comment]
How is it that this PR closes #11461? This PR creates approval rule associations, whereas #11461 asks for approval rule templates.

**Comment by zhelding** [comment]
Pull request #21306 has significantly refactored the AWS Provider codebase. As a result, most PRs opened prior to the refactor now have merge conflicts that must be resolved before proceeding.

Specifically, PR #21306 relocated the code for all AWS resources and data sources from a single `aws` directory to a large number of separate directories in `internal/service`, each corresponding to a particular AWS service. This separation of code has also allowed for us to simplify the names of underlying functions -- while still avoiding namespace collisions.

We recognize that many pull requests have been open for some time without yet being addressed by our maintainers. Therefore, we want to make it clear that resolving these conflicts _in no way affects the prioritization_ of a particular pull request. Once a pull request has been prioritized for review, the necessary changes will be made by a maintainer -- either directly or in collaboration with the pull request author.

For a more complete description of this refactor, including examples of how old filepaths and function names correspond to their new counterparts: please refer to issue #20000.

For a quick guide on how to amend your pull request to resolve the merge conflicts resulting from this refactor and bring it in line with our new code patterns: please refer to our [Service Package Refactor Pull Request Guide](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/service-package-pullrequest-guide.md).

**Comment by anGie44** [comment]
Hi again @carthewd , thank you for this PR! I'm going to be reviewing this PR and in the meantime, I recommend setting this PR to `allow maintainers to edit` so that we can help you update this PR for service packages (see #20000). If edits are possible, I will resolve merge conflicts on my end and get this ready for an upcoming release.

**Comment by anGie44** [comment]
Output of acceptance tests: (AWS Commercial):
```
--- PASS: TestAccCodeCommitApprovalRuleTemplate_disappears (15.48s)
--- PASS: TestAccCodeCommitApprovalRuleTemplateDataSource_basic (17.29s)
--- PASS: TestAccCodeCommitApprovalRuleTemplateAssociation_Disappears_repository (18.79s)
--- PASS: TestAccCodeCommitApprovalRuleTemplateAssociation_disappears (18.79s)
--- PASS: TestAccCodeCommitApprovalRuleTemplate_basic (19.37s)
--- PASS: TestAccCodeCommitApprovalRuleTemplateAssociation_basic (21.00s)
--- PASS: TestAccCodeCommitApprovalRuleTemplate_updateContentAndDescription (28.82s)
--- PASS: TestAccCodeCommitApprovalRuleTemplate_updateName (28.83s)
```

File: internal/service/codecommit/approval_rule_template_association.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	err = FindApprovalRuleTemplateAssociation(conn, approvalRuleTemplateName, repositoryName)
+
+	if !d.IsNewResource() && tfresource.NotFound(err) {
+		log.Printf("[WARN] CodeCommit Approval Rule Template Association (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("error reading Approval Rule Template Association (%s): %w", d.Id(), err)
```
Comment (anGie44) [inline internal/service/codecommit/approval_rule_template_association.go:87]:
```suggestion
		return fmt.Errorf("error reading CodeCommit Approval Rule Template Association (%s): %w", d.Id(), err)
```

**Review by anGie44 — APPROVED** [review]
Thanks again @carthewd for your contributions! Similar to #11487 , I've committed changes to fix merge conflicts that arose with the service packages restructure, and made updates to align with more recent contributing guidelines.

**Pull request merged**